include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp

```diff
@@ -242,20 +249,45 @@ inline typename seq_type::iterator erase_gap(seq_type & seq, typename seq_type::
  * the container. Before delegating, the function checks if the range
  * [\p first, \p last) contains only seqan3::gap symbols.
  */
-template <sequence_container_concept seq_type>
+template <sequence_container_concept gapped_seq_type>
 //!\cond
-    requires weakly_assignable_concept<reference_t<seq_type>, gap const &>
+    requires detail::is_gapped_alphabet<value_type_t<gapped_seq_type>>
 //!\endcond
-inline typename seq_type::iterator erase_gap(seq_type & seq,
-                                             typename seq_type::const_iterator first,
-                                             typename seq_type::const_iterator last)
+inline typename gapped_seq_type::iterator erase_gap(gapped_seq_type & seq,
+                                                    typename gapped_seq_type::const_iterator first,
+                                                    typename gapped_seq_type::const_iterator last)
 {
     for (auto it = first; it != last; ++it)
         if (*it != gap{}) // [[unlikely]]
             throw gap_erase_failure("The range to be erased contains at least one non-gap character.");
 
     return seq.erase(first, last);
 }
+
+/*!\brief An implementation of seqan3::aligned_sequence_concept::assign_ungapped_sequence for sequence containers.
+ * \tparam        gapped_seq_type   Type of the container to reassign; must model seqan3::sequence_container_concept;
+ *                                  the value type must be a seqan3::gapped alphabet.
+ * \tparam        ungapped_seq_type Type of the container to assign from; must model std::ranges::InputRange;
+ * \param[in,out] gapped            The gapped sequence container to assign to.
+ * \param[in,out] ungapped          The ungapped sequence container to assign from.
+ *
+ * \relates seqan3::aligned_sequence_concept
+ *
+ * \details
+ *
+ * This function clears the content of the `gapped` container and reassigns the content of the `ungapped` container by
+ * using std::copy.
+ */
+template <sequence_container_concept gapped_seq_type, std::ranges::InputRange ungapped_sequence_type>
+//!\cond
+    requires detail::is_gapped_alphabet<value_type_t<gapped_seq_type>> &&
+             weakly_assignable_concept<reference_t<gapped_seq_type>, reference_t<ungapped_sequence_type>>
+//!\endcond
+void assign_ungapped_sequence(gapped_seq_type & gapped, ungapped_sequence_type && ungapped)
+{
+    gapped.resize(ungapped.size());
+    std::copy(begin(ungapped), end(ungapped), begin(gapped));
```

Comment:
```suggestion
    std::copy(seqan3::begin(ungapped), seqan3::end(ungapped), seqan3::begin(gapped));
```

Reply:
You can't rely on unqualified lookup here.

Reply:
why?

Reply:
Well during unqualified lookup the compiler will only look in the class (friend member function) and the namespace of the class for a possible candidate, but not in other namespaces. So if you rely on std::begin to be picked up it won't unless you declare `using std::begin` before in the function body. Then also the namespace std:: will be searched for a possible candidate. To remedy this additional `using ...` declaration the ranges library introduced function objects as customisation points instead of free functions. These "functors" will do the proper lookup for you, but you need to use the qualified name (in our case our alias is fine). Thus, begin for ranges from other namespaces (not seqan3) will be resolved correctly.

Reply:
We should add a point "frequently used stuff from std::ranges::" to our next strategy meeting. And I would also propose making at least the free functions for our alphabets be "Niebloids" (function objects that are not found by ADL, but that perform ADL look-up if not delegating to members).